### ISSUE TITLE
VTKjs/vue component integration/build - Trouble with vtkVolumeController (TypeError: _ComputeHistogram ...)

### ISSUE THREAD

**Issue by luizcga** [Kitware/vtk-js#1177]
I'm making a VUEjs VTKjs component to display and manipulate volumes (CT VTI files).

After struggling a bit with configuration (npm, webpack, vue cli etc.) I finally got it working with the latest version of all libraries (vue 2.6.1 / vtk.js 9.4.1). 

I'm trying to follow the example volumeViewer (https://kitware.github.io/vtk-js/examples/VolumeViewer.html) and was able to turn it to a component that loads a VTI file (using the example head-binary-zlib.vti and others) from my local development server.

But when I run the app a get the following error:
Uncaught (in promise) TypeError: _ComputeHistogram_worker__WEBPACK_IMPORTED_MODULE_3__.default is not a constructor
    at Object.publicAPI.setDataArray (index.js?6ee7:454)
    at Object.publicAPI.setupContent (index.js?0f3b:181)
    at createViewer (vol3d.vue?ee7e:197)
    at eval (vol3d.vue?ee7e:265)

Debugging I found the error is thrown in line 454 of vtk.js\Sources\Interaction\Widgets\PiecewiseGaussianWidget\index.js 
"const worker = new ComputeHistogramWorker();"

--------------

If I don't try to set up the [vtkVolumeController] and call [setRGBTransferFunction] and [setScalarOpacity] manually, the volume loads normally.

All needed modules (listed in the example) are imported normally in the component's initial script section. 

Is this a webpack/babel configuration problem?
Can anyone give me a hint or point me in the right direction so that I can try to solve this issue?
Thanks!

**Comment by Robinxbnie** [comment]
I guess it is still the webpack/babel configuration problem. Anyway you have moved further by successfully getting it works with latest version of all libraries. In my case, it seems the libraries always conflict with each other, some required me upgrade the dependency to newer version while some other required the same dependency downgrade to older version. @luizcga,  Have you met the same pitfall?

**Comment by jourdain** [comment]
For vue application you should use the [vtk.js chainWebpack.js](https://github.com/Kitware/vtk-js/blob/master/Utilities/config/chainWebpack.js) to define the rules that vtk.js needs. You can even see an [example here](https://github.com/Kitware/web-project-templates/blob/master/vue-vtkjs-template/vue.config.js) of its usage.

**Comment by luizcga** [comment]
> For vue application you should use the [vtk.js chainWebpack.js](https://github.com/Kitware/vtk-js/blob/master/Utilities/config/chainWebpack.js) to define the rules that vtk.js needs. You can even see an [example here](https://github.com/Kitware/web-project-templates/blob/master/vue-vtkjs-template/vue.config.js) of its usage.

Cool, it's working now!
I already was using chanWebpack.js to define a rule for the glsl loader (without it the project doesn't even compile), but I wasn't aware of this neat configuration file for vtk.js. 

Thanks a lot!

**Comment by jourdain** [comment]
We try to provide some streamline way for vtk.js to get integrated in various frameworks but with so many options it is hard to support all of them. 

Also we are lacking documentation to illustrate its integration in various framework even if we already have several projects doing across React and Vue.

If you feel like it, a documentation guide for an integration in one of those framework would be great. A guide in our documentation is just a markdown file in that [directory](https://github.com/Kitware/vtk-js/tree/master/Documentation/content/docs).  Then I can help to expose it via the menu.

**Comment by luizcga** [comment]
> I guess it is still the webpack/babel configuration problem. Anyway you have moved further by successfully getting it works with latest version of all libraries. In my case, it seems the libraries always conflict with each other, some required me upgrade the dependency to newer version while some other required the same dependency downgrade to older version. @luizcga, Have you met the same pitfall?

Hi @Robinxbnie, yes I had lots of conflicts and experimented with various versions of all of them (vue-cli2, vue@cli3, babel, vtk.js etc.) but in the end I focused in trying to get all new versions to work.

Like I said in my previous comment, I had to write a custom rule for glsl loaders in [vue.config.js] (because vue@cli doesn't use the [webpack.config.js] anymore) so that my project could compile, but it still was throwing the runtime error I described in my original post.

@jourdain's pointer was the missing piece in the configuration puzzle :)
Just to help anyone that's also trying to achieve this, the versions I have installed are (from my project's [package.json] file):

"dependencies": {
    "core-js": "^2.6.5",
    "vtk.js": "^9.4.1",
    "vue": "^2.6.10"
  },
  "devDependencies": {
    "@vue/cli-plugin-babel": "^3.10.0",
    "@vue/cli-plugin-eslint": "^3.10.0",
    "@vue/cli-service": "^3.10.0",
    "babel-eslint": "^10.0.1",
    "eslint": "^5.16.0",
    "eslint-plugin-vue": "^5.0.0",
    "kw-web-suite": "^8.0.0",
    "vue-template-compiler": "^2.6.10",
    "vuetify-loader": "^1.2.2"
  },

Cheers!

**Comment by luizcga** [comment]
> We try to provide some streamline way for vtk.js to get integrated in various frameworks but with so many options it is hard to support all of them.
> 
> Also we are lacking documentation to illustrate its integration in various framework even if we already have several projects doing across React and Vue.
> 
> If you feel like it, a documentation guide for an integration in one of those framework would be great. A guide in our documentation is just a markdown file in that [directory](https://github.com/Kitware/vtk-js/tree/master/Documentation/content/docs). Then I can help to expose it via the menu.

Cool, I'll try to came up with something more detailed and usefull to help others later this month...
Thanks again!

**Comment by jourdain** [comment]
For your project, since you use `vue-cli`, you don't need to bring `kw-web-suite`, they kind of have the same role as bringing all the build tools. `vue-cli` bring even more things than `kw-web-suite`.

**Comment by luizcga** [comment]
> For your project, since you use `vue-cli`, you don't need to bring `kw-web-suite`, they kind of have the same role as bringing all the build tools. `vue-cli` bring even more things than `kw-web-suite`.

I uninstalled `kw-web-soute` but when I tried running my project it gave me new compilation errors:

 ERROR  Failed to compile with 11 errors                                                                       
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: shader-loader
You may need to install it.
Failed to resolve loader: worker-loader
You may need to install it.

Suggestions?

**Comment by jourdain** [comment]
ok then you just need to bring in `worker-loader` and `shader-loader`

**Comment by luizcga** [comment]
> ok then you just need to bring in `worker-loader` and `shader-loader`

Got it, working smoothly now ... 
Updated dependencies (from `package.json`):

  "dependencies": {
    "core-js": "^2.6.5",
    "shader-loader": "^1.3.1",
    "vtk.js": "^9.4.1",
    "vue": "^2.6.10",
    "worker-loader": "^2.0.0"
  },
  "devDependencies": {
    "@vue/cli-plugin-babel": "^3.10.0",
    "@vue/cli-plugin-eslint": "^3.10.0",
    "@vue/cli-service": "^3.10.0",
    "babel-eslint": "^10.0.1",
    "eslint": "^5.16.0",
    "eslint-plugin-vue": "^5.0.0",
    "node-sass": "^4.9.0",
    "sass": "^1.17.4",
    "sass-loader": "^7.1.0",
    "vue-template-compiler": "^2.6.10",
  },

**Comment by luizcga** [comment]
One more thing if I may.
In the `volumeViewer` example, a `vtkVolumeController` is created and used.

My component does the same (the code is very similar to the example). But for some reason the div that contains the controller has no style.

The class `VolumeController-module_container_2mLM3` (I assume provided by webpack) representing the `.container` selector of the `vtk.js\Sources\Interaction\UI\VolumeController\VolumeController.module.css` is assigned to it, but it has no values (doesn't exist).
The effect is a hidden and scrambled controller, with no style lost in the HTML.

This is probably another webpack configuration problem ... any suggestions on how to solve it ?

**Comment by jourdain** [comment]
This is the [rule](https://github.com/Kitware/vtk-js/blob/master/Utilities/config/chainWebpack.js#L43-L57) that actually process that style file.
I would think it should work.

But you can look at how we convert that controller into a vue widget [here](https://github.com/Kitware/paraview-glance/tree/master/src/components/widgets/PiecewiseFunctionEditor).

**Comment by luizcga** [comment]
> This is the [rule](https://github.com/Kitware/vtk-js/blob/master/Utilities/config/chainWebpack.js#L43-L57) that actually process that style file.
> I would think it should work.
> 
> But you can look at how we convert that controller into a vue widget [here](https://github.com/Kitware/paraview-glance/tree/master/src/components/widgets/PiecewiseFunctionEditor).

My bad ... 
Trying to contain the `fullScreenRenderer` in a small area, I was hacking away passing the `rootContainer` as the `Body` of an `iFrame`.
The `container` style was correctly set by `webpack` by didn't have scope inside the `iFrame` :(

Now I see I can define a simple `Div` component as the root and specify the style of the `container` (or even the `container` itself) in the `fullScreenRenderer` constructor.

Everything working now... learning more and more.
Very nice lib and even nicer results!

Thanks a lot.

**Comment by Robinxbnie** [comment]
> For vue application you should use the [vtk.js chainWebpack.js](https://github.com/Kitware/vtk-js/blob/master/Utilities/config/chainWebpack.js) to define the rules that vtk.js needs. You can even see an [example here](https://github.com/Kitware/web-project-templates/blob/master/vue-vtkjs-template/vue.config.js) of its usage.

Thanks @jourdain . I used this magic config and seems all version conflicts problems gone. Now I met another wired one. The compile failed by saying:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This dependency was not found:
* vue in ./src/main.js, ./src/App.vue and 13 others
To install it, you can run: npm install --save vue
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
It is no longer a vtkjs related issue now. But seems still related to config. I have vue installed of course (Basically this is a vue app). But seems the config not able to find my vue in node-modules.

Is there some tricks to explicitly do something in config files to deal with this?

Thanks and regards,
Robin

**Comment by Robinxbnie** [comment]
@jourdain. Forget about the issue I mentioned above. I tried to re-create the new project based on my old codes and npm install. The error not happened. I don't know the root cause while I just move to the new folder and can move forward now. Thanks a lot again.

**Comment by luizcga** [comment]
> This is the [rule](https://github.com/Kitware/vtk-js/blob/master/Utilities/config/chainWebpack.js#L43-L57) that actually process that style file.
> I would think it should work.
> 
> But you can look at how we convert that controller into a vue widget [here](https://github.com/Kitware/paraview-glance/tree/master/src/components/widgets/PiecewiseFunctionEditor).

Hi @jourdain still trying to get things to work over here...
I'm having some trouble with `vtkVolumeController` and I thing is config related.

Everything is working fine, except for the icons of the interface ... Webpack is loading the literal path (ex: `/img/logo.cf9c70b2.svg`) instead of the `SVG` files.

I've checked the `vtk.js/Utilities/config/chainWebpack.js` file and there no rule for `SVG` files.

Tried installing `vue-svg-loader` and adding a rule in my `config.vue.js` file but it didn't work (and broke my other `SVG` files because I didn't import them as modules).

Any suggestions about what may be the problem?
Thanks!
